edrumulus_hardware_teensy.cpp

```diff
@@ -0,0 +1,170 @@
+    conv_time = SFCAdder + Averages*(BCT + LSTAdder + HSCAdder)
+              = 5 ADCK + 5 bus + 1 *(25 ADCK + 0 ADCK + 2 ADCK) = 32 ADCK + 5 bus
+              = 13/7.5MHz + 5/60MHz = 4.35uS per conversion
+    For 22 conversions, 22*4.35us = 95.7uS which is still under 125uS.
+
+    F_BUS is 60MHz and F_ADC is max allowed 24MHz in 12-bit mode and 12MHz in
+    16-bit mode to stay within specs.  F_ADC is derived from F_BUS with dividers
+    1, 2, 4, 8, or 16.  F_BUS of 60MHz offers 30, 15, and 7.5 MHz options.
+
+    For 12-bit mode, 15 and 7.5 can work
+    For 16-bit mode, only 7.5 can work.
+
+    The definition for F_BUS can be found in kinetics.h and it's a function of
+    the F_CPU.  For F_CPU=180MHz, F_BUS=60MHz.  For F_CPU=192MHz, F_BUS=48MHz.
+    In other words, overclocking the teensy will affect the conversion rate.
+*/
+#endif
+
+  // initialize timer flag (semaphore)
+  timer_ready = false;
+
+  // prepare timer at a rate of given sampling rate
+  myTimer.begin ( on_timer, 1000000 / Fs ); // here we define the sampling rate (1 MHz / Fs)
+}
+
+
+void Edrumulus_hardware::on_timer()
+{
+  // tell the main loop that a sample can be read by setting the flag (semaphore)
+  edrumulus_hardware_pointer->timer_ready = true;
+}
+
+
+void Edrumulus_hardware::capture_samples ( const int number_pads,
+                                           const int number_inputs[],
+                                           int       analog_pin[][MAX_NUM_PAD_INPUTS],
+                                           int       sample_org[][MAX_NUM_PAD_INPUTS] )
+{
+  // wait for the timer to get the correct sampling rate when reading the analog value
+  while ( !timer_ready ) delayMicroseconds ( 5 );
+  timer_ready = false; // it is important to reset the flag here
+
+  // read the ADC samples
+  for ( int i = 0; i < total_number_inputs; i++ )
+  {
+#if defined(ARDUINO_TEENSY40) || defined(ARDUINO_TEENSY41) // Teensy 4.0/4.1 specific code
+    // pins 12 and 13 are ADC1 only, pins 10 and 11 are ADC0 only
+    // note that pin 8 gave large spikes on ADC0 but seems to work ok with ADC1
+    if ( ( input_pin[i] == 8 ) || ( input_pin[i] == 12 ) || ( input_pin[i] == 13 ) )
+    {
+      input_sample[i] = adc_obj.adc1->analogRead ( input_pin[i] );
+    }
+#elif defined(ARDUINO_TEENSY36)  // Teensy 3.6 specific code
+    // pins 12 and 13 and 22 are ADC1 only
+    if ( ( input_pin[i] == 12 ) || ( input_pin[i] == 13 ) || ( input_pin[i] == 22 ) )
+    {
+      input_sample[i] = adc_obj.adc1->analogRead ( input_pin[i] );
+    }
+#endif
+    else
+    {
+      input_sample[i] = adc_obj.adc0->analogRead ( input_pin[i] );
+    }
+  }
+
+  // copy captured samples in pad buffer
+  int input_cnt = 0;
+
+  for ( int i = 0; i < number_pads; i++ )
+  {
+    for ( int j = 0; j < number_inputs[i]; j++ )
+    {
+      sample_org[i][j] = input_sample[input_cnt++];
+#if defined(ARDUINO_TEENSY36)  // Teensy 3.6 specific code
+      sample_org[i][i] >>= 4;  // 16-bit to 12-bit hack
```

Comment:
You should not waste bits here. Increase the "pad_settings.velocity_threshold" value instead.

Reply:
If the bits 'wasted' are just noise then there's no issue.  Furthermore, adjusting the pad_settings for different adc bit widths is not good either.  Somehow the 16-bit and 12-bit readings need to be normalized so the settings look the same.

Reply:
> Furthermore, adjusting the pad_settings for different adc bit widths is not good either

I think I did not make myself clear. What I meant was that there are defines for different hardware properties and the signal processing changes on these defines so that the pad settings can be the same for each hardware. For the Teensy 3.6 I have now created the defines: https://github.com/corrados/edrumulus/blob/2a356179c0c5fcaec91a7a1c4ce7abb1e516f27c/edrumulus_hardware.h#L46

> sample_org[i][i] >>= 5;  // 16-bit to 12-bit hack

That means that you have 5 bits noise which is 2^5=32. I have assumed that the max. noise amplitude is about 20 (see my macro definition: https://github.com/corrados/edrumulus/blob/2a356179c0c5fcaec91a7a1c4ce7abb1e516f27c/edrumulus_hardware.h#L47)

Right now the threshold is defined as: https://github.com/corrados/edrumulus/blob/2a356179c0c5fcaec91a7a1c4ce7abb1e516f27c/edrumulus.cpp#L441. In this equation, the value of the macro "ADC_MAX_NOISE_AMPL" might be included to automatically adjust the threshold according to the given maximum noise amplitude of the ADC.